examples/dft.masm

```diff
@@ -0,0 +1,288 @@
+# A program which computes the Discrete Fourier Transform (DFT) of the provided public inputs.
+# We only allow input sizes of a power of 2 for convenience of computing the primitive root of unity.
+# This program makes use of the fact that the modulus for the Miden VM's prime field is 2^64 - 2^32 + 1.
+#
+# See the corresponding Wikipedia article: https://en.wikipedia.org/wiki/Discrete_Fourier_transform_over_a_ring
+#
+# Expects the input stack to be
+# Input: [<length of list>, <ele 0>, <ele 1>, ...]
+# 
+# Ouputs the Discrete Fourier Transform of the original list, which is a list of the same length.
+# Output: [<ele 0'>, <ele 1'>, ...]
+#
+# Convention: inputs are always consumed by the callee
+
+# Root of unity for domain of 2^32 elements, taken from Winterfell. See:
+# https://github.com/facebook/winterfell/blob/4543689f73a2efb9d30927535de7b4efe7e1802d/math/src/field/f64/mod.rs#L270
+const.TWO_ADIC_ROOT_OF_UNITY=7277203076849721926
+
+# This means there's a multiplicative subgroup of size 2^32 that exists in the field.
+# For more information about two-adicity, see https://cryptologie.net/article/559/whats-two-adicity/.
+const.TWO_ADICITY=32
+
+# Location where we will store the input array.
+# [0, 2^30) is global memory, so we store it at the beginning.
+const.INPUT_ARR_LOC=0
+
+# Location at 3 * 2^30, the first address "with no special meaning" in the root context
+const.RESULT_ARR_LOC=3221225472
+
+#! Computes log base 2. Only supports {1, 2, 4, 8, 16}.
+#! Input: [n]
+#! Output: [log2(n)] if n is supported, otherwise halts
+proc.log2
```

Comment:
This is cool, but basically you hardcode the result of the log2(n) operation, right? 

We could also compute the base-2 logarithm (log2) of an integer "n" and use a simple algorithm that repeatedly divides "n" by 2 until "n" becomes 1, while keeping track of the number of divisions. The number of divisions will be equal to the log2 of "n". Basically we can only compute the log2 of even numbers `<2^{32}`. 

```
proc.log2    
    # it should fail if the provided number >= 2^{32}
    u32test assert
    
    # if should faul if the provided number is odd
    dup push.2 u32checked_mod assertz

    # now we know that the number is even and < 2^{32}
    # so we can have a while loop and divide by 2 until we reach 1
    
    # create a counter
    push.0
    # get the loop going
    push.1

    while.true
        swap u32checked_div.2
        # get the counter and add + 1
        swap add.1 swap
        # check if the result != 1, then we continue in the loop
        dup movup.2 swap neq.1
    end

end
```

I am sure this code can be optimized.

Reply:
thinking about it, it's not 100% accurate. We don't need to check whether the number is odd or even, we would need to check if it is a power of 2 to ensure we get integers back

Reply:
Or we should only check if the provided number is a power of 2. And we can do that with bitwise comparison